src/core/server/saved_objects/migrations/model/helpers.ts

```diff
@@ -68,6 +69,55 @@ export function indexBelongsToLaterVersion(indexName: string, kibanaVersion: str
   return version != null ? gt(version, kibanaVersion) : false;
 }
 
+export function addMustNotClausesToQuery(
+  query: QueryDslQueryContainer,
+  mustNotClauses: QueryDslQueryContainer[]
+): QueryDslQueryContainer {
+  let filter: QueryDslQueryContainer[] | QueryDslQueryContainer = [];
+
+  if (query.bool?.filter) {
+    if (Array.isArray(query.bool.filter)) {
+      filter = filter.concat(query.bool.filter);
+
+      filter.push({
+        bool: {
+          must_not: mustNotClauses,
+        },
+      });
+    } else {
+      let currentMustNot: QueryDslQueryContainer[] = [];
+
+      if (query.bool.filter.bool?.must_not) {
+        currentMustNot = currentMustNot.concat(query.bool.filter.bool.must_not);
+      }
+
+      currentMustNot.push(...mustNotClauses);
+
+      filter = {
+        ...query.bool.filter,
+        bool: {
+          ...query.bool.filter.bool,
+          must_not: currentMustNot,
+        },
+      };
+    }
+  } else {
+    filter = {
+      bool: {
+        must_not: mustNotClauses,
+      },
+    };
+  }
+
+  return {
+    ...query,
+    bool: {
+      ...query.bool,
+      filter,
+    },
+  };
+}
```

Comment:
NOTE: the `QueryDslQueryContainer` type is awkward. All properties are declared optional, but in practice you can't actually have a container containing multiple entries, e.g `{ bool: boolQuery, match: matchQuery }` is invalid (unless I'm wrong, but I never saw such syntax anywhere in ES queries).

So our function only works for `bool` queries. Which is fine, given we're only using such type when we're using this util, but we may want to throw at the beginning of the function if `query.bool` is not defined.

Reply:
Makes sense, perhaps we could even consider having the input query as a `QueryDslBoolQuery` in that case.
Even the `excludeOnUpgradeQuery` could be a bool query.

Reply:
Yea, may be better if that doesn't impact too much the calling code